### ISSUE TITLE
Clarify HNAP interaction of main language and multiple PT_LOCALE  translations

### ISSUE THREAD

**Issue by jodygarnett** [metadata101/iso19139.ca.HNAP#12]
The following has been provided by NrCAN FGP in discussion with @ianwallen:

> The problem is that our geonetwork base is 3.8 and FGP geonetwork base is 3.6.  In geonetwork 3.8 they applied a change to the language setting. 
> So prior to 3.8 if the languages were English(main) and French then they would have a English as the main language and French would be in the PT_LOCALE. So in this case there was only one PT_locale.
> In 3.8, they still have English as the main language but they also have 2 PT_LOCALE – one for each language.  The multiple PT_Locale is causing the issue.
> I have included 2 version of the metadata that were attempted to be loaded.  The 3.6 version loads correctly.  The 3.8 version causes the issues – you test loading them if you want.
> 
> So we either need to apply a transformation prior to uploading the data to FGP so that it remove the main language PT_locale.  Or the FGP system is updated to support multiple PT_LOCALE. I believe it may be easier for us to apply a transformation prior to submitting the data to FGP.
> 
> I have already applied a bunch of fixes related to this issue in the new metadata 101 for HNAP (https://github.com/metadata101/iso19139.ca.HNAP)  for the 3.8 branch - so I presume that someday FGP will be upgraded to support this.

**Comment by jodygarnett** [comment]
The specification has the following example:

> Locale (PT_Locale)
>
> Guideline: NAP registered code lists based on ISO 639.2 alpha-3 terminology codes for language (eng=English, fra=French) and ISO 3166-1 alpha-3 codes (CAN=Canada) for countries shall be used to describe locale and character encoding shall be set to "utf8; utf8", as shown in the following example:
> 
>
> ```xml
>   <gmd:locale>
>     <gmd:PT_Locale id="fra">
>       <gmd:languageCode>
>         <gmd:LanguageCode codeListValue="fra" > codeList="http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml#IC_116">French; Français</gmd:LanguageCode>
>       </gmd:languageCode>
>       <gmd:country>
>         <gmd:Country codeListValue="CAN" codeList="http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml#IC_117" >Canada; Canada</gmd:Country>
>       </gmd:country>
>       <gmd:characterEncoding>
>         <gmd:MD_CharacterSetCode codeListValue="RI_458" codeList="http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml#IC_95" >utf8; utf8</gmd:MD_CharacterSetCode>
>       </gmd:characterEncoding>
>     </gmd:PT_Locale>
>   </gmd:locale>
> ```

This example only has the alternative language.

**Comment by ianwallen** [comment]
We have created a simple transformation stylesheet that will convert 3.8 to 3.6.  This transformation basically removes the primary language from the xml file.  We tested it and and we are able to load the transformed data to the FGP.   We currently use it as part of the publishing to FGP which is currently a custom solution (and a work in progress).  I can provide this transformation file if needed.

**Comment by jodygarnett** [comment]
Thanks Ian, I would like to better understand what is intended.

**Comment by ianwallen** [comment]
We are using our local geonetwork 3.8 installation to create/edit metadata records.  We intend to publish our approved metadata records to the FGP (Geonetwork 3.6) system (if it is geospatial data) and publish it to Open Data if it is none geospatial.  We may also publish to other 3rd parties.  We are currently using the Geonetwork api's for creating an MEF file and pushing the content to FGP.  When pushing our version to FGP, the data would be transferred but when we checked the data on the FGP system it was not being displayed correctly (missing most of the data). Once we applied the  transformation it corrected the issue.  I believe other departments may also have similar requirements to publish their metadata to FGP.  Currently the pushed data is showing up as a draft - we need to discuss with the FGP developers the possibility of exposing their publication module via api.

**Comment by ianwallen** [comment]
Similar to the extract-ckan-dataset.xsl which is used to publishing to Open Data, this is the xsl that we use to publish to FGP.  The submitted MEF files to FGP has the transformation applied.

[extract-fgp-dataset.zip](https://github.com/metadata101/iso19139.ca.HNAP/files/4345039/extract-fgp-dataset.zip)

**Comment by josegar74** [comment]
Ian, related to the main language added also as locale and duplication of the main language value in PT_FreeText elements, this was added in some 3.x version of Geonetwork, in previous versions it worked in a similar way as the FGP branch.

It can be disabled (that was done for FGP) as seem that is required for HNAP specification, although the HNAP specification doesn't really have any full metadata example, just snippets.

Have you check that the metadata created by DFO will be valid according to the HNAP specification?

**Comment by ianwallen** [comment]
Jose,
I just checked [#2032](https://github.com/geonetwork/core-geonetwork/pull/2032) again and looks like it may have been implemented in 3.4.

It is odd because, we had used the 3.6 napec schema plugin in 3.6 and it was working somewhat correctly (there were some java code changes as well as other change) with the geonetwork core 3.6. But when we upgraded to geonetwork core 3.8 there were lots of errors which were attributed to the main language not being displayed correctly. So I applied a bunch of fixes to make it work.

My understanding is that [#2032](https://github.com/geonetwork/core-geonetwork/pull/2032) was for making easier to work with the multilingual data as it only needs to worry about the textGroup.
Also it may be possible that in 3.8 more code was changed to only use the textGroup which would explain why the main language only became an issue in 3.8.
I believe it is easier in the long run to be compatible with the geonetwork-core than the attempt to disable/change the code.

As for being HNAP compliant, as you mention there deos not seem to be much documentation other than the xml schema.  The multilingual changes validate against the xml schema so in my opinion these changes are valid.  Looking at [#2032](https://github.com/geonetwork/core-geonetwork/pull/2032), placing the language in the textGroup and the CharacterString was a compromises that should be backwards compatible with ISO 19139 which HNAP is based on.
